### PR TITLE
[Dynamic] M2 for cache_read/write (S2)

### PR DESCRIPTION
**Pull request by spectrometerHBH**

### CODE REVIEW

**Comment by jinhongyii** [comment]
Can you give some documentations on what you fix in this PR?

File: tests/python/relax/test_dyn_cache_read.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+            T.writes(C[v0, v1])
+            C[v0, v1] = C_pad[v0, v1]
+
+
+def test():
+    sch = tir.Schedule(main, debug_mask="all")
+    b0 = sch.get_block(name="matmul", func_name="main")
+    b34 = sch.cache_read(block=b0, read_buffer_index=0, storage_scope="shared")
+    tvm.ir.assert_structural_equal(sch.mod["main"], expected)
+
```
Comment (jinhongyii) [inline tests/python/relax/test_dyn_cache_read.py:122]: please add test for cache write

File: src/tir/schedule/primitive/cache_read_write.cc [hunk 2]
```diff
@@ -1516,6 +1565,15 @@ StmtSRef CacheRead(ScheduleState self, const StmtSRef& block_sref, int read_buff
   }
 
   // Step 4. Making new cache stage block and rewrite readers.
+  info.cache_region = cache_region;
```
Comment (jinhongyii) [inline src/tir/schedule/primitive/cache_read_write.cc:1614]: just curious what if there's thread binding in outer scope
if we have threadIdx.x, it needs to be relaxed and the cache shape should be larger

Comment (spectrometerHBH) [reply src/tir/schedule/primitive/cache_read_write.cc:1614]: `AnalyzeRegionUpperBound` should have taken this into account.

**Pull request merged**